### PR TITLE
Adaptive refactorization

### PR DESCRIPTION
**Pull request by mmuckley**
A draft for refactorizations for #205 to reduce duplicate code.

### CODE REVIEW

**Comment by Timsey** [comment]
On a first pass:

Possibly major stuff:
- EquispacedMaskFunc is gone from subsample.py and instead we have EquiSpacedMaskFunc and EquispacedMaskFractionFunc. EquispacedMaskFractionFunc seems to have the same functionality as EquispacedMaskFunc from the adaptive VarNet code, but it is missing a block controlled by the `skip_low_freqs` argument (we used this to create masks that had the proper number of lines when combined with dense low-frequency sampling): https://github.com/facebookresearch/fastMRI/blob/58c0bfcd676630cc3f002a8129a1388dc8896307/fastmri/data/subsample.py#L212-L233

Minor stuff:
- Comment here is now a lie and we can just remove it: https://github.com/facebookresearch/fastMRI/blob/69db5c4e49ed0eda93bae8ed4daefacd6b2235b5/fastmri/models/active_varnet.py#L145
- Argparse description for `val_volume_sample_rate` and `test_volume_sample_rate` should refer to `val_sample_rate` and `test_sample_rate` (and vice-versa), not to `sample_rate`: https://github.com/facebookresearch/fastMRI/blob/69db5c4e49ed0eda93bae8ed4daefacd6b2235b5/fastmri/pl_modules/data_module.py#L370
- Argument `active_acquisition` in MiniCoilTransform is unused and can be removed everywhere in that class: https://github.com/facebookresearch/fastMRI/blob/69db5c4e49ed0eda93bae8ed4daefacd6b2235b5/fastmri/data/transforms.py#L525
- Example folder is now called _active_varnet_, but should be called _adaptive_varnet_. Same for active_varnet.py and active_varnet_module.py.
- MiniCoilTransform and VarNetTransform now also return full kspace. This is necessary for sign-fixing in the adaptive VarNet, but not for the base VarNet, which does now also receive kspace in the forward(). Not necessarily bad, but I just thought I'd mention it as a potential thing to clean up.

**Comment by mmuckley** [comment]
@Timsey could you explain the first a bit more? Do you just want the acceleration mask where the densely-sampled center does not exist?

**Comment by Timsey** [comment]
> @Timsey could you explain the first a bit more? Do you just want the acceleration mask where the densely-sampled center does not exist?

So the original EquispacedMaskFunc that I used (now EquispacedMaskFractionFunc) does not fully take into account the densely sampled center when computing `accel_samples` here: https://github.com/facebookresearch/fastMRI/blob/mmuckley/adaptive_varnet_updates/fastmri/data/subsample.py#L337 This does compute an adjusted acceleration, but it can still end up sampling one or more of its lines in the center, leading to overlapping samples and varying actual acceleration factors (based on e.g. offset). 

The quoted code block ensures a constant acceleration by forcing the equisipaced sampling to happen outside of the low-frequency region. The downside of this is that samples on one side of the low-frequency region are often not exactly equispaced w.r.t. samples on the other side of the low-frequency region.

File: fastmri/data/transforms.py [hunk 1]
```diff
@@ -509,3 +509,150 @@ def __call__(
...
+        self.crop_size = crop_size
+        self.num_compressed_coils = num_compressed_coils
+        self.active_acquistion = active_acquisition
+
+    def __call__(self, kspace, mask, target, attrs, fname, slice_num):
+        """
+        Args:
+            kspace: Input k-space of shape (num_coils, rows, cols) for
+                multi-coil data.
+            mask: Mask from the test dataset.
```
Comment (luisenp) [inline fastmri/data/transforms.py:552]: Looks like this is not used when `self.mask_func` is given. If this is the case, can we add an additional sentence here explaining this?

**Comment by mmuckley** [comment]
Sorry I should have commented here: @Timsey I added your version of the `EquispacedMaskFunc` to the `fastmri_examples` folder for your project.

**Review by luisenp — APPROVED** [review]
LGTM! Left some minor comments/questions.

File: fastmri/data/transforms.py [hunk 2]
```diff
@@ -509,3 +509,150 @@ def __call__(
...
+    Multi-coil compressed transform, for faster prototyping.
+    """
+
+    def __init__(
+        self,
+        mask_func: Optional[MaskFunc] = None,
+        use_seed: Optional[bool] = True,
+        crop_size: Optional[tuple] = None,
+        num_compressed_coils: Optional[int] = None,
+        active_acquisition: Optional[bool] = False,
```
[the remark this answers is not in the document]
Comment (mmuckley) [reply fastmri/data/transforms.py]: Looks like it can be removed. I just took it out.

**Comment by Timsey** [comment]
LGTM! One more thing I noticed is that with MriModule reverted to the original, a bunch of the stuff we log with wandb is missing now, so I would propose just removing that logging. 

@luisenp We should probably also add that single-checkpoint evaluation script?

**Comment by mmuckley** [comment]
@luisenp @Timsey We do have single-checkpoint eval sketches here if it helps, but you all can use whatever format you prefer.

https://github.com/facebookresearch/fastMRI/blob/main/fastmri_examples/varnet/run_pretrained_varnet_inference.py

File: fastmri/data/transforms.py [hunk 3]
```diff
@@ -509,3 +509,147 @@ def __call__(
...
+        left_vec, _, _ = np.linalg.svd(kspace, compute_uv=True, full_matrices=False)
+        kspace = np.reshape(
+            np.array(np.matrix(left_vec[:, :num_compressed_coils]).H @ kspace),
+            new_shape,
+        )
+        kspace = to_tensor(kspace)
+
+        # Mask kspace
+        if self.mask_func:
+            masked_kspace, mask = apply_mask(
```
Comment (luisenp) [inline fastmri/data/transforms.py:627]: I get an error here when trying to run the first command in the README, since `apply_mask` also returns a third element with `num_low_frequencies`.

File: fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    ent[prob_mask == 0] = 0
+    ent[prob_mask == 1] = 0
+    return ent
+
+
+def save_images(recon):
+    for i, tensor in enumerate(recon):
+        plt.imshow(tensor.detach().cpu().numpy(), cmap="gist_gray")
+        plt.colorbar()
+        plt.savefig(f"/scratch/lep/active_acq/recon_{i}.png")
```
Comment (Timsey) [inline fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py:38]: Should we make this save path an arg, or alternatively save to some parent of the checkpoint dir?

Comment (Timsey) [inline fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py:38]: Should we make this save path an arg, or alternatively save to some parent of the checkpoint dir?

File: fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        sample_rate=args.sample_rate,
+        volume_sample_rate=args.volume_sample_rate,
+        batch_size=args.batch_size,
+        num_workers=args.num_workers,
+    )
+
+    # ------------
+    # model
+    # ------------
+    device = "cuda:0"
```
Comment (Timsey) [inline fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py:79]: Probably fine like this, just pointing our this hardcoded bit in case we want to change it.

File: fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py (hunk 4 above)
Comment (luisenp) [reply fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py:38]: Oh, that's my bad, I missed this. I will add an --output_dir to args.

File: fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        else data_module.train_dataloader()
+    )
+    with torch.no_grad():
+        for i, batch in enumerate(data_loader):
+            if i == args.num_batches:
+                break
+            kspace, masked_kspace, mask, target, fname, slice_num, max_value, _ = batch
+
+            # import pdb
+            # pdb.set_trace()
```
Comment (Timsey) [inline fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py:99]: Remove commented code in line 98 and 99.

File: fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+                kspace.to(device),
+                masked_kspace.to(device),
+                mask.to(device),
+                fname,
+                slice_num,
+            )
+            prob_masks_list = extra_outputs["prob_masks"]
+
+            # import pdb
+            # pdb.set_trace()
```
Comment (Timsey) [inline fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py:110]: Remove commented code in lines 109 and 110.

File: fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py (hunk 4 above)
Comment (luisenp) [reply fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py:38]: Actually, this function is not even called anymore, so might as well remove it. Not sure about including it in the main code, since presumably saving all images in the validation dataset is not something the user wants to do.

File: fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+            output, extra_outputs = model(
+                kspace.to(device),
+                masked_kspace.to(device),
+                mask.to(device),
+                fname,
+                slice_num,
+            )
+            prob_masks_list = extra_outputs["prob_masks"]
```
Comment (Timsey) [inline fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py:107]: This or lines below will probably give an error for Equispaced models (no prob_mask). We already have an `if all_prob_masks` check in line 156, so we should add a similar check here (`if "prob_masks" in extra_output:`).

File: fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py (hunk 4 above)
Comment (Timsey) [reply fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py]: Ah, that makes sense! Probably not necessary indeed, and not a great default setting anyway since it takes a lot of disk space.

File: fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py (hunk 8 above)
Comment (luisenp) [reply fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py:97]: I think in its current version this script doesn't support Equispaced models at all. Is there a way to automatically detect the type of model from the checkpoint? In my internal scripts I used to do it from the wandb id, but I don't think we should rely on that here.

Comment (Timsey) [reply fastmri_examples/adaptive_varnet/eval_pretrained_adaptive_varnet.py:96]: Aah yes, good point. If we want to make this work for equispaced models we probably have to do a try-except on the checkpoint loading with ActiveVarNetModule and VarNetModule. In the latter case we'd also need to check if mask type is `equispaced` or `random` if we want to be complete.

The eval script also assumes (128, 128) crop and 4 compressed coils, whereas the train scripts leaves this as an argument (and if no coil-compression is used, it uses VarNetDataTransform instead: https://github.com/facebookresearch/fastMRI/blob/mmuckley/adaptive_varnet_updates/fastmri_examples/adaptive_varnet/train_adaptive_varnet_demo.py#L410 ).

I don't really see a problem with leaving the eval script as is, but these changes would make it more compatible with the full optionality that the training script provides.

**Comment by Timsey** [comment]
Random comment: this line here can be simplified, and the comment removed: https://github.com/facebookresearch/fastMRI/blob/mmuckley/adaptive_varnet_updates/fastmri_examples/adaptive_varnet/subsample.py#L128. Simplification is just to remove the entire line and change `skip` here to `num_low_frequencies`: https://github.com/facebookresearch/fastMRI/blob/mmuckley/adaptive_varnet_updates/fastmri_examples/adaptive_varnet/subsample.py#L134.

**Comment by Timsey** [comment]
Another issue, this one a bit larger.

This bit of code will crash if we ever reach it: 
https://github.com/facebookresearch/fastMRI/blob/mmuckley/adaptive_varnet_updates/fastmri_examples/adaptive_varnet/train_adaptive_varnet_demo.py#L411-L415

because `VarNetDataTransform` returns a `VarNetSample`:
https://github.com/facebookresearch/fastMRI/blob/mmuckley/adaptive_varnet_updates/fastmri/data/transforms.py#L500-L509

but `AdaptiveVarNetModule` expects a `tuple` as returned by `MiniCoilTransform`:
https://github.com/facebookresearch/fastMRI/blob/mmuckley/adaptive_varnet_updates/fastmri/data/transforms.py#L646-L655

see:
https://github.com/facebookresearch/fastMRI/blob/mmuckley/adaptive_varnet_updates/fastmri/pl_modules/adaptive_varnet_module.py#L197

This also creates an issue for training equispaced models with our training script, because `VarNetModule` expects a `VarNetSample`:
https://github.com/facebookresearch/fastMRI/blob/mmuckley/adaptive_varnet_updates/fastmri/pl_modules/adaptive_varnet_module.py#L197

So we either have to:
1) Change MiniCoilTransform and AdaptiveVarNetModule to accept VarNetSample (note that this VarNetSample also needs to contain the full kspace for sign-fixing purposes). 
2) Make our own copy `VarNetDataTransform` that returns a tuple like `MiniCoilTransform`, and our own copy of `VarNetModule` that expects a tuple like `AdaptiveVarNetModule`.

**Comment by Timsey** [comment]
> So we either have to:
> 
> 1. Change `MiniCoilTransform` and `AdaptiveVarNetModule` to accept `VarNetSample` (note that this `VarNetSample` also needs to contain the full `kspace` for sign-fixing purposes).
> 2. Make our own copy `VarNetDataTransform` that returns a tuple like `MiniCoilTransform`, and our own copy of `VarNetModule` that expects a tuple like `AdaptiveVarNetModule`.

I'm thinking option 2 is better now, since I'm also running into some issues with `VarNetModule` not expecting any of the later-added args, such as `num_sense_lines`.

**Comment by mmuckley** [comment]
@Timsey I can spend some time on these. How are you launching jobs locally?
